After Signin Modal

sbsmith86 opened the pull request:
#### What's this PR do?

This PR does a few things:
- Sets up a configuration option on Custom Settings that allow admins to check if organ donation should be allowed. [here](https://github.com/DoSomething/phoenix/compare/after-signin-modal?diff=unified&expand=1&name=after-signin-modal#diff-1d14f80a019bcc7b992a2faa28d8f707R511)
- If that ^^ flag is turned on, after a user signs up the modal is displayed and `Takeover.js` injects content into the modal and handles everything from there. But we also check that the users hasn't yet submitted the registration form before deciding if we should show it or not. [here](https://github.com/DoSomething/phoenix/compare/after-signin-modal?diff=unified&expand=1&name=after-signin-modal#diff-ae398bb8864fc3c26317f7ccaef04c0aR765)
- in `Takeover.js`, this PR, starts to roughly mock out how the modal will work for organ donations. We pull in some underscore templates that house the content for each "screen" in the modal. Flow is like:
  - User clicks the submit button on the first "screen" 
  - We make an AJAX call to a fake endpoint. 
  - If that is successful, we move to the next screen. 
  - User hits the submit button there, an AJAX called is made to a custom endpoint in phoenix that stores data about the registration. see #6491 
#### How should this be reviewed?

Pull branch down. Make sure you have the `dosomething_organ_donation` module turned on. Go to a campaign and turn on organ donation in the custom settings form. 

Now, sign up for the campaign (preferably as a new user or anon user since that is what most of the folks will be) when you get taken to the action page, you should see the modal. And the experience should work like what is outlined above.
#### Any background context you want to provide?

This is not final at all, but covers the ability to show a modal after a user signs up for a campaign and allows us to start making API calls from within that modal. [Please see my notes here](https://github.com/DoSomething/phoenix/compare/after-signin-modal?diff=unified&expand=1&name=after-signin-modal#diff-18be182c71ce2870f0f2020d6740e029R189) about some of the JS and how I would like to refactor it. 
#### Relevant tickets

Fixes #6481 

@angaither @DFurnes @weerd @chloealee

sbsmith86 committed:
Add a configuration option for adding an organ donation registration experience

sbsmith86 committed:
Show the modal on page load

sbsmith86 committed:
Set up JS, SCSS files

sbsmith86 committed:
Show form on first modal screen, move to next screen after successful ajax request

sbsmith86 committed:
Move to the next screen in the modal after a successful call to an API

sbsmith86 committed:
Add validation handling to zipcode field

sbsmith86 committed:
Adds back button

sbsmith86 committed:
Consolidate functions that handle moving between screen
Start roughly mocking out the different screens we need

sbsmith86 committed:
Move ajax request handling into a general function that takes in requirements for the request
Pass sid to the front end using the drupal.settings object
Store registration in phoenix db when submit button is clicked.

sbsmith86 committed:
Add check if the user already submitted the registration form, if so, it doesn't show

sbsmith86 committed:
Add warning comment about WIP code

sbsmith86 commented on lib/modules/dosomething/dosomething_campaign/dosomething_campaign.admin.inc:
```diff
@@ -499,3 +502,56 @@ function dosomething_campaign_archive_activity_form_submit(&$form, &$form_state)
+ * organ donation modal after signup.
+ *
+ */
+function dosomething_campaign_organ_donation_config_form($form, &$form_state, $node) {
```
Now that i'm looking at it, i feel like this admin config stuff should be housed in the `dosomething_organ_donation` module. But it's not a blocker for merge, I don't think...

angaither commented on lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:
```diff
@@ -669,6 +669,9 @@ function dosomething_campaign_node_view($node, $view_mode, $langcode) {
     dosomething_campaign_add_signup_data_form_vars($node);
 
+    // Flag if we need to register an organ donor on the page.
+    dosomething_campaign_add_oragan_donation_vars($node);
```
I feel confused by this function call and the name.

This checks to see if we should add the organ donation vars? Shouldn't we say

```
if (dosomething_campaign_organ_donation_enabled)
  // add vars
  dosomething_campaign_add_organ_donation_vars($node);
```

angaither commented on lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:
```diff
@@ -752,6 +755,23 @@ function dosomething_campaign_add_signup_data_form_vars(&$node) {
+function dosomething_campaign_add_oragan_donation_vars(&$node) {
+  $organ_registration_turned_on = dosomething_helpers_get_variable('node', $node->nid, 'register_organ_donors');
+
+  $sid = dosomething_signup_exists($node->nid);
```
this `sid` and `timestamp` check should be a helper function 
`dosomething_campaign_organ_doner_has_registered`

mshmsh5000 replied:
I am a jerk, but let's please make sure we're spelling `organ` and `donor` correctly everywhere: http://cl.ly/3B1c0t3r1g2K

sbsmith86 replied:
@mshmsh5000 :facepalm: major typo on my part

DFurnes commented on lib/modules/dosomething/dosomething_campaign/dosomething_campaign.admin.inc:
```diff
@@ -499,3 +502,56 @@ function dosomething_campaign_archive_activity_form_submit(&$form, &$form_state)
+  // Load the node's helpers variables.
+  $vars = dosomething_helpers_get_variables('node', $node->nid);
+
+  $form['nid'] = array(
```
It'd be lovely to start using array literal syntax here `[ ... ]`!

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/takeover/Takeover.js:
```diff
@@ -0,0 +1,245 @@
+const template = require('lodash/string/template');
+
+const Takeover = {
+  baseUrl: $(location).attr("origin"),
```
Is there an advantage to wrapping `location` in a jQuery element?

If not, I think `window.location.origin` might be more clear.

sbsmith86 replied:
other than it allows you to do `.attr()`, nothing. I will update.

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/takeover/Takeover.js:
```diff
@@ -0,0 +1,245 @@
+    const _this = this;
+
+    // Create custom deferred promise object.
+    let ajaxPromise = $.Deferred().resolve().promise();
```
Definitely not something needed for merge, but it'd be pretty cool to start using [fetch](https://jakearchibald.com/2015/thats-so-fetch/) via [the polyfill](https://github.com/github/fetch) since the way jQuery handles promises is so wonky.

sbsmith86 replied:
Yea i saw you talking about this, but haven't spent the time to look into it...maybe when I refactor stuff.

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/takeover/Takeover.js:
```diff
@@ -0,0 +1,245 @@
+        ajaxPromise = ajaxPromise.then(function() {
+          return promise;
+        }).done(function(data) {
+          _this.changeScreens('next');
```
I've started using `=>` arrow functions lately to be able to keep `this` without having to save it in a temporary variable. Totally a stylistic decision, but I feel like it makes things read a bit nicer (plus you can make a lot of things one-liners!)

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/js/takeover/Takeover.js:
```diff
@@ -0,0 +1,245 @@
+    // Check if the new screen exists, if so, go there, otherwise we need to loop back.
+    if ($.inArray(newIndex, this.screenIndexes)) {
+      $current.removeClass('active');
+      $('.takeover__screen[data-index="' + newIndex + '"]').addClass('active');
```
This'd be a lovely place to use a [template string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals):

``` js
$(`.takeover__screen[data-index="${newIndex}"]`)
```

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/scss/content/_takeover.scss:
```diff
@@ -0,0 +1,85 @@
+  height: 600px;
+  overflow: hidden;
+
+  &:after {
```
You can use the `@include clearfix;` mix-in for this!

DFurnes commented on lib/themes/dosomething/paraneue_dosomething/scss/content/_takeover.scss:
```diff
@@ -0,0 +1,85 @@
+    transition: $transition;
+    max-width: 1350px;
+
+    &.active {
```
Let's use `.is-active` if this is a "state" class.

sbsmith86 committed:
Move organ donation admin stuff into the organ donation module
Refactor how we check if someone has registered so it is more readable

sbsmith86 committed:
Syntax updates to the JS for better JS

sbsmith86 commented:
@angaither I moved the all the admin stuff for the organ donation configuration into the `dosomething_organ_donation` module and then added that helper you mentioned there as well. Then just checked if the module exists and if the user has register to determine how to set the variable. I hope the makes it a little more readable!

@DFurnes thanks for your comments! i didn't know about `=>` syntax and it is lovely! I also addressed your other comments and will work on this further, refactoring to classes like we discussed IRL

sbsmith86 committed:
remove all the storing of 'this'

angaither commented:
👍
